ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -367,194 +365,205 @@ let maybe_get_guest_metrics ~__context ~ref =
+     let current_error = check current_error (fun () ->
+         if op = `checkpoint || op = `snapshot || op = `suspend || op = `snapshot_with_quiesce
+         then (* If any vdi exists with on_boot=reset, then disallow checkpoint, snapshot, suspend *)
+           if List.exists fst !!vdis_reset_and_caching
+           then Some (Api_errors.vdi_on_boot_mode_incompatible_with_operation,[])
+           else None
+         else if op = `pool_migrate then
+           (* If any vdi exists with on_boot=reset and caching is enabled, disallow migrate *)
+           if List.exists (fun (reset,caching) -> reset && caching) !!vdis_reset_and_caching
+           then Some (Api_errors.vdi_on_boot_mode_incompatible_with_operation,[])
+           else None
+         else None) in
+
+     (* If a PCI device is passed-through, check if the operation is allowed *)
+     let current_error = check current_error (fun () ->
+         if vmr.Db_actions.vM_attached_PCIs <> []
+         then check_pci ~op ~ref_str
+         else None) in
+
+     (* The VM has a VGPU, check if the operation is allowed*)
+     let current_error = check current_error (fun () ->
+         if vmr.Db_actions.vM_VGPUs <> []
+         then check_vgpu ~__context ~op ~ref_str ~vgpus:vmr.Db_actions.vM_VGPUs
+         else None) in
+
+     (* Check for errors caused by VM being in an appliance. *)
+     let current_error = check current_error (fun () ->
+         if is_appliance
+         then check_appliance ~vmr ~op ~ref_str
+         else None) in
+
+     (* Check for errors caused by VM being assigned to a protection policy. *)
+     let current_error = check current_error (fun () ->
+         if has_protection_policy
+         then check_protection_policy ~vmr ~op ~ref_str
+         else None) in
+
+     (* Check for errors caused by VM being assigned to a snapshot schedule. *)
+     let current_error = check current_error (fun () ->
+         if has_snapshot_schedule
+         then check_snapshot_schedule ~vmr ~ref_str op
+         else None) in
+
+     (* Check whether this VM needs to be a system domain. *)
+     let current_error = check current_error (fun () ->
+         if op = `query_services && not (bool_of_assoc "is_system_domain" vmr.Db_actions.vM_other_config)
+         then Some (Api_errors.not_system_domain, [ ref_str ])
+         else None) in
+
+     (* For storage migration, check if any of the VM's VDIs has CBT enabled *)
+     let current_error = check current_error (fun () ->
+         let is_cbt_enabled vdi = try Db.VDI.get_cbt_enabled ~__context ~self:vdi with _ -> false in
+         if op = `migrate_send && List.exists is_cbt_enabled !!vdis then begin
+           let vdi = List.find is_cbt_enabled !!vdis in
+           Some (Api_errors.vdi_cbt_enabled, [ Ref.string_of vdi ])
+         end else None
+       ) in
+
+     let current_error = check current_error (fun () ->
+         if Helpers.rolling_upgrade_in_progress ~__context &&
```

Comment:
This also does a `Db.Pool.get_other_config`, so we could move this too out of the loop, but I guess that's not a big deal.

Reply:
thanks for spotting that, I think we  could also swap the arguments of && to shortcircuit evaluation